### PR TITLE
additional_data: prometheus: Add location lookup metric

### PR DESCRIPTION
**Pull request by R2ZER0**
Add a metric to help us monitor when our geocode lookup info goes out of date.

### CODE REVIEW

File: datastore/prometheus/views.py [hunk 1]
```diff
@@ -85,13 +99,74 @@ def _total_num_sources_in_last_run(self):
             NUM_PROBLEM_SOURCES_IN_LAST_RUN.set(-1)
             NUM_OK_SOURCES_IN_LAST_RUN.set(0)
 
+    def _num_current_grants_with_beneficiary_location_geocode_without_lookup(self):
+        current_grants = db.Latest.objects.get(series=db.Latest.CURRENT).grant_set
+
```
Comment (michaelwood) [inline datastore/prometheus/views.py:104]: Didn't someone add a convince function this ? ;) 

https://github.com/ThreeSixtyGiving/datastore/blob/master/datastore/db/models.py#L22

File: datastore/prometheus/views.py [hunk 2]
```diff
@@ -85,13 +99,74 @@ def _total_num_sources_in_last_run(self):
...
+        # Workaround, fixed in Django 5.0
+        # See: https://stackoverflow.com/questions/42332304/django-an-equality-check-in-annotate-clause
+        class GreaterThan(Func):
+            arg_joiner = ">"
+            arity = 2
+            function = ""
+
+        zero = Cast(0, output_field=IntegerField())
+
+        # This query is a bit gnarley using postgres-specific JSON functions
```
Comment (michaelwood) [inline datastore/prometheus/views.py:119]: No need to change this now but if it becomes more complex than its worth (readability, maintainability) there is prior precedent for falling back to using [raw sql](https://github.com/ThreeSixtyGiving/datastore/blob/master/datastore/data_quality/quality_data.py#L258) , or [partial](https://github.com/ThreeSixtyGiving/datastore/blob/master/datastore/data_quality/quality_data.py#L236)  [etc](https://github.com/ThreeSixtyGiving/datastore/blob/master/datastore/db/management/commands/manage_entities_data.py#L77).

**Review by michaelwood — COMMENTED** [review]
Looks ok to me, I'm a little worried that the complexity of this means it could do with some error handling to avoid a monitoring failure.. Happy to be reassured though.

File: datastore/prometheus/views.py (hunk 1 above)
Comment (R2ZER0) [reply datastore/prometheus/views.py:104]: huh.... oh yeah! :D

File: datastore/prometheus/views.py [hunk 3]
```diff
@@ -85,13 +99,74 @@ def _total_num_sources_in_last_run(self):
...
+        )
+
     def get(self, *args, **kwargs):
         # Update gauges unless we're in the middle of processing/loading
         if db.Status.all_idle_and_ready():
             self._num_errors_log()
             self._total_latest_grants()
             self._total_datagetter_grants()
             self._total_num_sources_in_last_run()
+            self._num_current_grants_with_beneficiary_location_geocode_without_lookup()
```
[the remark this answers is not in the document]
Comment (R2ZER0) [reply datastore/prometheus/views.py:169]: No wait, i misunderstood how timeit works. 13.5 seconds was the combined time for all 10 runs.
Trying a few individual runs, it's taken 1.5-1.9 seconds, which is better!

The SQL query generated is:
```sql
SELECT 
  "db_grant"."id", 
  "db_grant"."grant_id", 
  "db_grant"."data", 
  "db_grant"."getter_run_id", 
  "db_grant"."publisher_id", 
  "db_grant"."source_file_id", 
  "db_grant"."additional_data", 
  jsonb_path_query_array(
    "db_grant"."data", 
    $.beneficiaryLocation[ * ] ? (
      exists(@.geoCode)
    )
  ) AS "geocodes", 
  jsonb_array_length(
    (
      "db_grant"."additional_data" -> locationLookup
    )
  ) AS "lookup_count", 
  jsonb_array_length(
    jsonb_path_query_array(
      "db_grant"."data", 
      $.beneficiaryLocation[ * ] ? (
        exists(@.geoCode)
      )
    )
  ) AS "geocode_count", 
  (
    CAST(
      jsonb_array_length(
        jsonb_path_query_array(
          "db_grant"."data", 
          $.beneficiaryLocation[ * ] ? (
            exists(@.geoCode)
          )
        )
      ) AS integer
    )> CAST(0 AS integer)
  ) AS "has_geocode", 
  (
    CAST(
      jsonb_array_length(
        (
          "db_grant"."additional_data" -> locationLookup
        )
      ) AS integer
    )> CAST(0 AS integer)
  ) AS "has_lookup" 
FROM 
  "db_grant" 
  INNER JOIN "db_grant_latest" ON (
    "db_grant"."id" = "db_grant_latest"."grant_id"
  ) 
WHERE 
  (
    "db_grant_latest"."latest_id" = 1575 
    AND (
      CAST(
        jsonb_array_length(
          jsonb_path_query_array(
            "db_grant"."data", 
            $.beneficiaryLocation[ * ] ? (
              exists(@.geoCode)
            )
          )
        ) AS integer
      )> CAST(0 AS integer)
    ) 
    AND NOT (
      CAST(
        jsonb_array_length(
          (
            "db_grant"."additional_data" -> locationLookup
          )
        ) AS integer
      )> CAST(0 AS integer)
    )
  )

```

which is about as readable as could be expected :) At least it does show nothing too unusual is happening, it's a simple select on the grants table.

**Pull request merged**